src/Sarif/SarifUtilities.cs

```diff
@@ -207,5 +208,22 @@ public static int GetByteLength(string s, Encoding encoding)
             byte[] bytes = encoding.GetBytes(s);
             return bytes.Length;
         }
+
+        internal static bool UnitTesting = false;
+
+        internal static void DebugAssert(bool conditional)
```

Comment:
>DebugAssert [](start = 29, length = 11)

Random thing I'm throwing out there. Debug.Assert statements are great, except that they aren't handled well while unit testing (the tests resolve as inconclusive). This pattern would converter a failed assert into an exception, assuming tests set the UnitTesting conditional. Note that this would also allow us to effectively inject the assert into retail builds (Debug.Assert will compile away entirely of course). This is a nice value, unit testing can provoke some assertion coverage in release testing that would otherwise never get executed.

Want to start a general conversation on this, as we'll all want to be in sync on this approach assuming we proceed.

Reply:
I like the idea of getting better unit test results. I suggest adding a `message` or `reason` string argument. You can use it as the message argument to `InvalidOperartionException` and as the second argument to `Debug.Assert`.

---
In reply to: [244507788](https://github.com/Microsoft/sarif-sdk/pull/1191#discussion_r244507788) [](ancestors = 244507788)

Reply:
A fantastic idea. 

---
In reply to: [244542519](https://github.com/Microsoft/sarif-sdk/pull/1191#discussion_r244542519) [](ancestors = 244542519,244507788)